core/unit_test/hip/TestHIP_SharedResourceLocking.cpp

```diff
@@ -0,0 +1,84 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <random>
+
+#include <Kokkos_Core.hpp>
+#include <TestHIP_Category.hpp>
+
+struct ThreadWorkOnSharedResource {
+  int device;
+  size_t value;
+  size_t *counter;
+
+  // Detached work that ends up releasing the shared resource lock.
+  void callback() const {
+    std::this_thread::sleep_for(std::chrono::microseconds{value});
+    *counter += value;
+    Kokkos::Impl::HIPInternal::constantMemReusable[device].release();
+  }
+
+  // This is what is typically done in HIPParallelLaunchKernelInvoker::invoke_kernel.
+  // First, acquire the lock, then spawn some asynchronous work that will release the shared resource lock.
+  void operator()() const {
+    [[maybe_unused]]const auto lock = Kokkos::Impl::HIPInternal::constantMemReusable[device].acquire();
+    std::thread(&ThreadWorkOnSharedResource::callback, *this).detach();
+  }
+};
+
+
+TEST(TEST_CATEGORY, shared_resource_locking) {
+  constexpr size_t nthreads_per_device = 2<<5;
```

Comment:
@dalg24 This test just ensures that all the machinery works. On our machine it takes only a few milliseconds.

Commenting the `acquire` and `release` makes it fail because non-atomic accumulation occurs with data races.